added int8 support for negate kernel

lamarrr opened the pull request:
removed the todo message, and plugged in the default **reference_ops**'s negate.

lamarrr committed:
added int8 support for negate kernel

removed the todo message, and plugged in the default reference_ops's negate kernel.

lamarrr committed:
added switch break

my bad, my tests revealed I forgot to add a break statement for the type deduction switch

lamarrr committed:
implemented unit tests for quantized int8 neg

I added unit tests for the quantized (int8) version of the neg op

renjie-liu commented on tensorflow/lite/micro/kernels/neg.cc:
```diff
@@ -31,15 +31,22 @@ TfLiteStatus Eval(TfLiteContext* context, TfLiteNode* node) {
   switch (input->type) {
-    // TODO(wangtz): handle for kTfLiteInt8
+    case kTfLiteInt8:
+      reference_ops::Negate(GetTensorShape(input), GetTensorData<int8_t>(input),
```
it seems this only works when zero_point == 0.

can you handle other cases as well?

thanks

lamarrr replied:
Works in what sense?
My assumption was that irregardless of the input being quantized or not, negation is agnostic of the zero-point.
Is this assumption flawed?

renjie-liu commented:
A simple example:

Let's say your zero_point is 6 and your scale is 1.5.

And your current int8 value is 8, so your float value is float_value =
(quantized_value - zero_point) * scale which is (8 - 6) * 1.5 = 3.

The negate value of that one should be -3.

In your implementation which is a direct negate, so 8 -> -8, so the result
float value will be (-8 - 6) * 1.5 = -21, which is way off -3.

Does that make sense?

On Mon, Jan 6, 2020 at 10:44 AM Basit Ayantunde <notifications@github.com>
wrote:

> *@lamarrr* commented on this pull request.
> ------------------------------
>
> In tensorflow/lite/micro/kernels/neg.cc
> <https://github.com/tensorflow/tensorflow/pull/35461#discussion_r363141413>
> :
>
> > @@ -31,15 +31,22 @@ TfLiteStatus Eval(TfLiteContext* context, TfLiteNode* node) {
>    const TfLiteTensor* input = GetInput(context, node, kInputTensor);
>    TfLiteTensor* output = GetOutput(context, node, kOutputTensor);
>    switch (input->type) {
> -    // TODO(wangtz): handle for kTfLiteInt8
> +    case kTfLiteInt8:
> +      reference_ops::Negate(GetTensorShape(input), GetTensorData<int8_t>(input),
>
> Works in what sense?
> My assumption was that irregardless of the input being quantized or not,
> negation is agnostic of the zero-point.
> Is this assumption flawed?
>
> —
> You are receiving this because your review was requested.
> Reply to this email directly, view it on GitHub
> <https://github.com/tensorflow/tensorflow/pull/35461?email_source=notifications&email_token=AIURNGORG6DJ73UP72OPMZLQ4KLHFA5CNFSM4KALN4O2YY3PNVWWK3TUL52HS4DFWFIHK3DMKJSXC5LFON2FEZLWNFSXPKTDN5WW2ZLOORPWSZGOCQWE3SA#discussion_r363141413>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AIURNGNGSMA7L74WF5TWEBTQ4KLHFANCNFSM4KALN4OQ>
> .
>


-- 
Renjie Liu

renjieliu@google.com
+1 (650) 253-4359

lamarrr commented:
> A simple example: Let's say your zero_point is 6 and your scale is 1.5. And your current int8 value is 8, so your float value is float_value = (quantized_value - zero_point) * scale which is (8 - 6) * 1.5 = 3. The negate value of that one should be -3. In your implementation which is a direct negate, so 8 -> -8, so the result float value will be (-8 - 6) * 1.5 = -21, which is way off -3. Does that make sense?
> […](#)
> On Mon, Jan 6, 2020 at 10:44 AM Basit Ayantunde ***@***.***> wrote: ***@***.**** commented on this pull request. ------------------------------ In tensorflow/lite/micro/kernels/neg.cc <[#35461 (comment)](https://github.com/tensorflow/tensorflow/pull/35461#discussion_r363141413)> : > @@ -31,15 +31,22 @@ TfLiteStatus Eval(TfLiteContext* context, TfLiteNode* node) { const TfLiteTensor* input = GetInput(context, node, kInputTensor); TfLiteTensor* output = GetOutput(context, node, kOutputTensor); switch (input->type) { - // TODO(wangtz): handle for kTfLiteInt8 + case kTfLiteInt8: + reference_ops::Negate(GetTensorShape(input), GetTensorData<int8_t>(input), Works in what sense? My assumption was that irregardless of the input being quantized or not, negation is agnostic of the zero-point. Is this assumption flawed? — You are receiving this because your review was requested. Reply to this email directly, view it on GitHub <#35461?email_source=notifications&email_token=AIURNGORG6DJ73UP72OPMZLQ4KLHFA5CNFSM4KALN4O2YY3PNVWWK3TUL52HS4DFWFIHK3DMKJSXC5LFON2FEZLWNFSXPKTDN5WW2ZLOORPWSZGOCQWE3SA#discussion_r363141413>, or unsubscribe <https://github.com/notifications/unsubscribe-auth/AIURNGNGSMA7L74WF5TWEBTQ4KLHFANCNFSM4KALN4OQ> .
> -- Renjie Liu renjieliu@google.com +1 (650) 253-4359

Makes perfect sense, thanks so much for the heads-up! 

So, now we'll definitely need to perform a multiplication and subtraction on the int8 rep.
How do I handle this? Should I accumulate on int16 and saturate to int8 (max +127) ?

renjie-liu commented:
Given it's negate: where output = -input, so I think we should make sure the output_scale == input_scale, but note the input_zeropoint maybe different from output_zeropoint.

So we have:
input_float = (input_q - input_zp) * input_scale
and
output_float = (output_q - output_zp) * output_scale

since output_float = -input_float, we have:

(output_q - output_zp) * output_scale = -(input_q - input_zp) * input_scale

Since output_scale == input_scale,  we can get:

output_q - output_zp = input_zp - input_q

which is:

output_q = (input_zp + output_zp) - input_q

Note input_zp + output_zp is a constant, so we can pre-compute it, you will just end up with

output_q = constant - input_q

and yes, you will need to saturate the result to make sure it's within the range.

Hope this helps.

lamarrr commented:
> Given it's negate: where output = -input, so I think we should make sure the output_scale == input_scale, but note the input_zeropoint maybe different from output_zeropoint.
> 
> So we have:
> input_float = (input_q - input_zp) * input_scale
> and
> output_float = (output_q - output_zp) * output_scale
> 
> since output_float = -input_float, we have:
> 
> (output_q - output_zp) * output_scale = -(input_q - input_zp) * input_scale
> 
> Since output_scale == input_scale, we can get:
> 
> output_q - output_zp = input_zp - input_q
> 
> which is:
> 
> output_q = (input_zp + output_zp) - input_q
> 
> Note input_zp + output_zp is a constant, so we can pre-compute it, you will just end up with
> 
> output_q = constant - input_q
> 
> and yes, you will need to saturate the result to make sure it's within the range.
> 
> Hope this helps.

Awesome!
I already derived the equation using different scales, I suppose yours is a specialization of that.
I will be pushing updates from my local branch here soon. Thanks so much.

lamarrr committed:
Merge pull request #4 from lamarrr/master

post 2.1 changes

lamarrr committed:
added support for varying zero points

I added the quantized int8 negation implementation for input & output with same scale and/not different zero points. I also added the neg op to the reference_base in the BUILD file so that it will be visible

lamarrr commented:
Hello @renjie-liu,
I have made the changes, thanks for your feedback.

lamarrr commented:
I also added the op to the reference integer_ops instead of using the default op (for float).

renjie-liu commented on tensorflow/lite/kernels/internal/reference/integer_ops/neg.h:
```diff
@@ -0,0 +1,64 @@
+  const int flat_size = MatchingFlatSize(input_shape, output_shape);
+
+  // already within int8 range, stored in int32
+  const auto prior = static_cast<int16_t>(input_zero_point + output_zero_point);
```
this is a constant, so we can pre-compute this in the prepare function

lamarrr replied:
Yes, It is only computed once. 
The pattern I used is the same in other kernels (The Non-quantized `Negate` especially). 
Also, there is no `Prepare` function for `Register_NEG's` `TfLiteRegistration`.
Are you sure you still want me to change this? (Create `Prepare` function)

renjie-liu commented on tensorflow/lite/micro/kernels/neg_test.cc:
```diff
@@ -71,6 +71,72 @@ void TestNegFloat(std::initializer_list<int> input_dims_data,
+  constexpr int tensors_size = inputs_size + outputs_size;
+
+  // quantize input
+  std::transform(input, input + element_count, quantized_input_data,
```
can you add the test in tensorflow/lite/kernels as well (also the kernel).

FYI, micro and tflite shares the most kernel implementation.

renjie-liu commented:
The prepare function is here:
https://github.com/tensorflow/tensorflow/blob/master/tensorflow/lite/kernels/neg.cc#L31-L40

On Wed, Jan 15, 2020 at 7:19 AM Basit Ayantunde <notifications@github.com>
wrote:

> *@lamarrr* commented on this pull request.
> ------------------------------
>
> In tensorflow/lite/kernels/internal/reference/integer_ops/neg.h
> <https://github.com/tensorflow/tensorflow/pull/35461#discussion_r366624355>
> :
>
> > +      static_cast<int16_t>(std::numeric_limits<int8_t>::max());
> +  constexpr int16_t kI8Min =
> +      static_cast<int16_t>(std::numeric_limits<int8_t>::min());
> +
> +  // within: [-128, 127]
> +  TFLITE_DCHECK_GE(input_zero_point, static_cast<int32_t>(kI8Min));
> +  TFLITE_DCHECK_LE(input_zero_point, static_cast<int32_t>(kI8Max));
> +
> +  // within: [-128, 127]
> +  TFLITE_DCHECK_GE(output_zero_point, static_cast<int32_t>(kI8Min));
> +  TFLITE_DCHECK_LE(output_zero_point, static_cast<int32_t>(kI8Max));
> +
> +  const int flat_size = MatchingFlatSize(input_shape, output_shape);
> +
> +  // already within int8 range, stored in int32
> +  const auto prior = static_cast<int16_t>(input_zero_point + output_zero_point);
>
> Yes, It is only computed once.
> The pattern I used is the same in other kernels (The Non-quantized Negate
> especially).
> Also, there is no Prepare function for Register_NEG's TfLiteRegistration.
> Are you sure you still want me to change this? (Create Prepare function)
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/tensorflow/tensorflow/pull/35461?email_source=notifications&email_token=AIURNGL5R376VU4FTF5A673Q5ZCBZA5CNFSM4KALN4O2YY3PNVWWK3TUL52HS4DFWFIHK3DMKJSXC5LFON2FEZLWNFSXPKTDN5WW2ZLOORPWSZGOCRYEXKA#discussion_r366624355>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AIURNGKEOSFOULQ3K62DS6LQ5ZCBZANCNFSM4KALN4OQ>
> .
>


-- 
Renjie Liu

renjieliu@google.com
+1 (650) 253-4359

lamarrr commented:
> The prepare function is here: https://github.com/tensorflow/tensorflow/blob/master/tensorflow/lite/kernels/neg.cc#L31-L40
> […](#)
> On Wed, Jan 15, 2020 at 7:19 AM Basit Ayantunde ***@***.***> wrote: ***@***.**** commented on this pull request. ------------------------------ In tensorflow/lite/kernels/internal/reference/integer_ops/neg.h <[#35461 (comment)](https://github.com/tensorflow/tensorflow/pull/35461#discussion_r366624355)> : > + static_cast<int16_t>(std::numeric_limits<int8_t>::max()); + constexpr int16_t kI8Min = + static_cast<int16_t>(std::numeric_limits<int8_t>::min()); + + // within: [-128, 127] + TFLITE_DCHECK_GE(input_zero_point, static_cast<int32_t>(kI8Min)); + TFLITE_DCHECK_LE(input_zero_point, static_cast<int32_t>(kI8Max)); + + // within: [-128, 127] + TFLITE_DCHECK_GE(output_zero_point, static_cast<int32_t>(kI8Min)); + TFLITE_DCHECK_LE(output_zero_point, static_cast<int32_t>(kI8Max)); + + const int flat_size = MatchingFlatSize(input_shape, output_shape); + + // already within int8 range, stored in int32 + const auto prior = static_cast<int16_t>(input_zero_point + output_zero_point); Yes, It is only computed once. The pattern I used is the same in other kernels (The Non-quantized Negate especially). Also, there is no Prepare function for Register_NEG's TfLiteRegistration. Are you sure you still want me to change this? (Create Prepare function) — You are receiving this because you were mentioned. Reply to this email directly, view it on GitHub <#35461?email_source=notifications&email_token=AIURNGL5R376VU4FTF5A673Q5ZCBZA5CNFSM4KALN4O2YY3PNVWWK3TUL52HS4DFWFIHK3DMKJSXC5LFON2FEZLWNFSXPKTDN5WW2ZLOORPWSZGOCRYEXKA#discussion_r366624355>, or unsubscribe <https://github.com/notifications/unsubscribe-auth/AIURNGKEOSFOULQ3K62DS6LQ5ZCBZANCNFSM4KALN4OQ> .
> -- Renjie Liu renjieliu@google.com +1 (650) 253-4359

Thanks. I was looking at tensorflow/lite/micro/kernels/neg.cc.
I'll make the necessary changes later today.

renjie-liu commented:
Oh, if you intend to change for micro as well. Feel free to add another
prepare function.

On Wed, Jan 15, 2020 at 10:00 AM Basit Ayantunde <notifications@github.com>
wrote:

> The prepare function is here:
> https://github.com/tensorflow/tensorflow/blob/master/tensorflow/lite/kernels/neg.cc#L31-L40
> … <#m_1576323117430476474_>
> On Wed, Jan 15, 2020 at 7:19 AM Basit Ayantunde *@*.*> wrote: @.**
> commented on this pull request. ------------------------------ In
> tensorflow/lite/kernels/internal/reference/integer_ops/neg.h <#35461
> (comment)
> <https://github.com/tensorflow/tensorflow/pull/35461#discussion_r366624355>>
> : > + static_cast<int16_t>(std::numeric_limits<int8_t>::max()); + constexpr
> int16_t kI8Min = +
> static_cast<int16_t>(std::numeric_limits<int8_t>::min()); + + // within:
> [-128, 127] + TFLITE_DCHECK_GE(input_zero_point,
> static_cast<int32_t>(kI8Min)); + TFLITE_DCHECK_LE(input_zero_point,
> static_cast<int32_t>(kI8Max)); + + // within: [-128, 127] +
> TFLITE_DCHECK_GE(output_zero_point, static_cast<int32_t>(kI8Min)); +
> TFLITE_DCHECK_LE(output_zero_point, static_cast<int32_t>(kI8Max)); + +
> const int flat_size = MatchingFlatSize(input_shape, output_shape); + + //
> already within int8 range, stored in int32 + const auto prior =
> static_cast<int16_t>(input_zero_point + output_zero_point); Yes, It is only
> computed once. The pattern I used is the same in other kernels (The
> Non-quantized Negate especially). Also, there is no Prepare function for
> Register_NEG's TfLiteRegistration. Are you sure you still want me to change
> this? (Create Prepare function) — You are receiving this because you were
> mentioned. Reply to this email directly, view it on GitHub <#35461
> <https://github.com/tensorflow/tensorflow/pull/35461>?email_source=notifications&email_token=AIURNGL5R376VU4FTF5A673Q5ZCBZA5CNFSM4KALN4O2YY3PNVWWK3TUL52HS4DFWFIHK3DMKJSXC5LFON2FEZLWNFSXPKTDN5WW2ZLOORPWSZGOCRYEXKA#discussion_r366624355>,
> or unsubscribe
> https://github.com/notifications/unsubscribe-auth/AIURNGKEOSFOULQ3K62DS6LQ5ZCBZANCNFSM4KALN4OQ
> .
> -- Renjie Liu renjieliu@google.com +1 (650) 253-4359 <(650)%20253-4359>
>
> Thanks. I was looking at tensorflow/lite/micro/kernels/neg.cc.
> I'll make the necessary changes later today.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/tensorflow/tensorflow/pull/35461?email_source=notifications&email_token=AIURNGI2CPACR6WKV7BQNZDQ5ZU4LA5CNFSM4KALN4O2YY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOEI6ZCBA#issuecomment-574460164>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AIURNGJFOPZTAUZ5UCTLVEDQ5ZU4LANCNFSM4KALN4OQ>
> .
>


-- 
Renjie Liu

renjieliu@google.com
+1 (650) 253-4359

lamarrr commented:
> The prepare function is here: https://github.com/tensorflow/tensorflow/blob/master/tensorflow/lite/kernels/neg.cc#L31-L40
> […](#)
> On Wed, Jan 15, 2020 at 7:19 AM Basit Ayantunde ***@***.***> wrote: ***@***.**** commented on this pull request. ------------------------------ In tensorflow/lite/kernels/internal/reference/integer_ops/neg.h <[#35461 (comment)](https://github.com/tensorflow/tensorflow/pull/35461#discussion_r366624355)> : > + static_cast<int16_t>(std::numeric_limits<int8_t>::max()); + constexpr int16_t kI8Min = + static_cast<int16_t>(std::numeric_limits<int8_t>::min()); + + // within: [-128, 127] + TFLITE_DCHECK_GE(input_zero_point, static_cast<int32_t>(kI8Min)); + TFLITE_DCHECK_LE(input_zero_point, static_cast<int32_t>(kI8Max)); + + // within: [-128, 127] + TFLITE_DCHECK_GE(output_zero_point, static_cast<int32_t>(kI8Min)); + TFLITE_DCHECK_LE(output_zero_point, static_cast<int32_t>(kI8Max)); + + const int flat_size = MatchingFlatSize(input_shape, output_shape); + + // already within int8 range, stored in int32 + const auto prior = static_cast<int16_t>(input_zero_point + output_zero_point); Yes, It is only computed once. The pattern I used is the same in other kernels (The Non-quantized Negate especially). Also, there is no Prepare function for Register_NEG's TfLiteRegistration. Are you sure you still want me to change this? (Create Prepare function) — You are receiving this because you were mentioned. Reply to this email directly, view it on GitHub <#35461?email_source=notifications&email_token=AIURNGL5R376VU4FTF5A673Q5ZCBZA5CNFSM4KALN4O2YY3PNVWWK3TUL52HS4DFWFIHK3DMKJSXC5LFON2FEZLWNFSXPKTDN5WW2ZLOORPWSZGOCRYEXKA#discussion_r366624355>, or unsubscribe <https://github.com/notifications/unsubscribe-auth/AIURNGKEOSFOULQ3K62DS6LQ5ZCBZANCNFSM4KALN4OQ> .
> -- Renjie Liu renjieliu@google.com +1 (650) 253-4359


**From my investigation**:

1.  The Prepare function only checks the nodes and sets the input sizes (setting the sizes of the inputs and isn't required since there is no dynamic tensor resizing; embedded targets).
2.  It does not seem `Prepare` is meant to be used for calculating the `OpData` (`prior`, in this case).
3.  The `OpData` needs to be calculated in the `EvalQuantizedInt8` function or in the `Op` itself. 
4.  I calculated it in the `Op` itself since it is just **one parameter** and having a separate `OpData` for it will be redundant.
5.  I do not know of any specific name for the sum of two `zero-points`. I feel this is an implementation-specific detail.
6.  It is only calculated once, just as it would when using the `OpData`, and prevents code duplication across the Tensorflow Lite and Tensorflow Lite Micro implementations.

lamarrr committed:
added Prepare function for TfLite Micro's Prepare

lamarrr committed:
added Quantized Int8 Neg implementation to TfLite's Neg Op

lamarrr committed:
added Quantized NegOpModel and Quantized Int8 Tests to TfLite

renjie-liu commented:
you should be able to find examples about how to precompute stuff in prepare, for example, see fully_connected (precompute multiplier): https://github.com/tensorflow/tensorflow/blob/master/tensorflow/lite/kernels/fully_connected.cc#L168-L178

also, for zero point, you can get from quantization params: https://github.com/tensorflow/tensorflow/blob/master/tensorflow/lite/c/common.h#L349

renjie-liu commented:
the prepare should be ran only once, the eval will be called every time the tflite interpreter got invoked. I'm fine if you find putting the pre-compute logic in prepare is tricky.

can but you add the kernel to tflite(not just micro) and test as well?

thanks!

lamarrr commented:
> you should be able to find examples about how to precompute stuff in prepare, for example, see fully_connected (precompute multiplier): https://github.com/tensorflow/tensorflow/blob/master/tensorflow/lite/kernels/fully_connected.cc#L168-L178
> 
> also, for zero point, you can get from quantization params: https://github.com/tensorflow/tensorflow/blob/master/tensorflow/lite/c/common.h#L349

Thanks, That only applies to TfLite, not TfLite Micro since you can't do dynamic memory allocation in the embedded context.

lamarrr commented:
> the prepare should be ran only once, the eval will be called every time the tflite interpreter got invoked. I'm fine if you find putting the pre-compute logic in prepare is tricky.
> 
> can but you add the kernel to tflite(not just micro) and test as well?
> 
> thanks!

Thanks for the clarification. I've already added the kernel and tests to TfLite.

lamarrr committed:
Removed Zero Points' Sum Computation and Cleaned Implementation

I removed the computation of the sum of the zero points, so it'll be handled by the Prepare/Eval functions as appropriate

lamarrr committed:
Moved Zero Point Sum Pre-computation to the Prepare Arena

lamarrr committed:
Added Zero Point Sum Computation to TfLite Micro's Eval

lamarrr commented:
@renjie-liu

renjie-liu commented on tensorflow/lite/kernels/neg_test.cc:
```diff
@@ -70,5 +70,48 @@ TEST(NegOpModel, NegInt64) {
+
+  EXPECT_THAT(
+      model.GetDequantizedOutput<int8_t>(),
+      ElementsAreArray(ArrayFloatNear(expected_output, quantized_tolerance)));
```
nit: since we can get the result pretty accurate, we don't need a large tolerance right?

lamarrr replied:
We do, the tolerance depends on the range being represented.
As we perform the calculation in the int8 space, there will be a quantization loss as the floating-point space won't be fully represented. Other kernels/Ops' tests use the same approach.

I also noticed a bug in my code and fixed it. Thanks!

renjie-liu commented on tensorflow/lite/kernels/internal/reference/integer_ops/neg.h:
```diff
@@ -0,0 +1,58 @@
+
+}  // namespace reference_integer_ops
+}  // namespace tflite
+#endif  // TENSORFLOW_LITE_KERNELS_INTERNAL_REFERENCE_INTEGER_OPS_NEG_H_
```
nit:please add a newline

lamarrr replied:
Ok, great!

lamarrr replied:
I've resolved it.
My linter funnily didn't warn me on that.
Thanks!

renjie-liu reviewed (commented):
looks really good, thanks!

renjie-liu commented:
adding Tiezhen to take a look at the micro kernel & test

lamarrr commented on tensorflow/lite/kernels/internal/reference/integer_ops/neg.h:
```diff
@@ -0,0 +1,58 @@
+  // = 382
+  // thus, accumulate on int16 [-383, 382]
+
+  constexpr auto kI8Min =
```
I was under the impression that it'd be verbose since we already specified it's type in the static_cast

lamarrr replied:
done

lamarrr commented on tensorflow/lite/kernels/neg.cc:
```diff
@@ -28,13 +29,58 @@ namespace neg {
 constexpr int kInputTensor = 0;
 constexpr int kOutputTensor = 0;
 
+struct OpDataInt8 {
```
The OpData's data type is int8 specific (Add and subtract on int16)

renjie-liu replied:
I mean the name should be OpData, int8 or uint8 or other datatypes can be just different fields in OpData. (it's easier for us to allocate one type)

wangtz replied:
+1 for OpData 

Init method can only allocate a single struct. In order to reduce binary and memory footprint, you'll need to create a int8 flavor of neg operator - which we don't have good facility to do so.

lamarrr replied:
Alright, I understand better now. Thanks for the clarification.
I've resolved it.

renjie-liu reviewed (commented):
looks good, please wait for Tiezhen's review on micro as well. thanks

wangtz reviewed (commented):
Please resolve the comments. otherwise looks good.

wangtz commented on tensorflow/lite/micro/kernels/neg_test.cc:
```diff
@@ -71,6 +71,72 @@ void TestNegFloat(std::initializer_list<int> input_dims_data,
+
+  // quantize input
+  std::transform(input, input + element_count, quantized_input_data,
+                 std::bind(tflite::testing::F2QS, std::placeholders::_1,
```
Is there a way to avoid using std::bind? We're trying to reduce as much template magic as possible in TFLM but I'll leave @petewarden to make the call.

lamarrr replied:
I've replaced it with a lambda instead

lamarrr committed:
Changed OpDataInt8 to OpData

I've changed OpDataInt8 (int8 specific) to OpData since we can easily allocate a single struct for all data types instead of different ones

lamarrr committed:
changed implicitly inferred type name to explicit

I changed the auto inferred type name to int16_t

lamarrr committed:
added new line to end of file

added new line to end of file to resolve linter warnings

lamarrr committed:
removed std::bind and placeholders

as suggested, I removed the templated std::bind and placeholders and replaced them with a lambda

lamarrr committed:
fixed int8 tolerance calculation

i noticed an implicit cast would occur with my int8 tolerance calculation, subtle since -6.0f, 6.0f (min, max to be represented) is same as -6 and 6 (int32) repectively. apparently the compiler flags weren't set to warn about this type of error

lamarrr commented:
> Please resolve the comments. otherwise looks good.

Done, thanks!

gbaned commented:
@lamarrr Can you please resolve conflicts? Thanks!

lamarrr committed:
Merge pull request #5 from tensorflow/master

pulled commits from Tensorflow:master

lamarrr committed:
added docstrings to indicate Registration struct's members

lamarrr committed:
fixed docstrings for TfLite's (and Micro's) Op Registration struct

lamarrr committed:
resolve conflicts

lamarrr committed:
Merge branch 'tensorflow-master' into u_negate_kernel

resolved change conflicts with incoming changes

lamarrr commented:
> @lamarrr Can you please resolve conflicts? Thanks!


Sorry for being late, been occupied with exams. 
Resolved! 
Thanks.

lamarrr commented:
@petewarden @renjie-liu

gbaned commented:
@lamarrr Still, conflicts are appearing. Can you please resolve? Thanks!

lamarrr committed:
Merge pull request #7 from tensorflow/master

Merge PRs from Tensorflow:master

lamarrr committed:
resolved merge conflicts

lamarrr committed:
Merge branch 'master' into u_negate_kernel

lamarrr commented:
> @lamarrr Still, conflicts are appearing. Can you please resolve? Thanks!

Yeah, another one appeared, The context->ReportError was changed to TF_LITE_KERNEL_LOG, seems that's how other kernels are. So I conformed with it.
I've resolved it now. Thanks again!

lamarrr commented:
> @lamarrr Still, conflicts are appearing. Can you please resolve? Thanks!

@gbaned @petewarden @renjie-liu

lamarrr commented:
@petewarden @gbaned

rthadur commented on tensorflow/lite/kernels/neg.cc:
```diff
@@ -58,10 +104,18 @@ TfLiteStatus Eval(TfLiteContext* context, TfLiteNode* node) {
+          GetTensorShape(input), GetTensorData<int8_t>(input),
+          GetTensorShape(output), GetTensorData<int8_t>(output),
+          op_data->zero_points_sum);
+    } break;
```
`third_party/tensorflow/lite/micro/kernels/neg.cc:113:54: error: missing field 'profiling_string' initializer [-Werror,-Wmissing-field-initializers]
                                 /*invoke=*/neg::Eval}`

rthadur replied:
please fix above error

lamarrr replied:
Fixed.
Also, -Wmissing-field-initializers is not enabled here.

lamarrr committed:
re-ordedered neg registration struct initialization

gbaned commented:
@lamarrr Can you please resolve conflicts? Thanks!

tensorflowbutler commented:
It has been 14 days with no activity and the `awaiting response` label was assigned. Is this PR still valid? Assigning the `stalled` label. Please comment to reassure me that this is still being worked on.

gbaned commented:
I'm going to go ahead and close this PR, because it seems to have stalled. If you're still interested in pursing this (and responding to my comments), please feel free to reopen!

lamarrr committed:
merge changes from master to child branch

lamarrr committed:
made negate kernel's log more descriptive

lamarrr committed:
fixed negate kernel's registration function prepare routine

lamarrr committed:
fixed negate kernel's op data calculate function

lamarrr committed:
replaced GetTensorShape and GetTensorData with tflu equivalent

lamarrr committed:
removed initializer lists in negate kernel's tests

lamarrr committed:
updated CreateQuantizedTensor to meet new api

lamarrr committed:
made negate kernel use the new kernelrunner

lamarrr commented:
Sorry about that. I didn't stall intentionally. I've had some business and issues over the past couple of months and had to be away from Open Source. I'm back now.

lamarrr commented:
I've resolved the conflict @gbaned

lamarrr committed:
updated includes

lamarrr committed:
Merge pull request #11 from tensorflow/master

merge master into child

lamarrr commented:
awaiting review @renjie-liu @petewarden

advaitjain commented:
@lamarrr thanks for your patience.

Given the long history of this PR and the recent changes to the [Micro contribution guidelines](https://github.com/tensorflow/tensorflow/blob/master/tensorflow/lite/micro/CONTRIBUTING.md), I am going to request (with apologies for all the delays) that we take a step back and follow the new guidelines.

My specific recommendation is:
 * Please make a github issue [motivating the new feature](https://github.com/tensorflow/tensorflow/blob/master/tensorflow/lite/micro/CONTRIBUTING.md) that you would like -- in an effort to give contributors faster feedback and PR reviews, we need to focus efforts on features that have broad impact. This will unfortunately mean saying no to some features.

If we agree that now is a good time for this feature, I would say the next steps would be:

 * Since there was a lot of back and forth with @renjie-liu write up the details of the int8 math in the github issue so that is is documented for future reference and can also be referred to during the PR review.

 * Break the work into smaller PRs that will be easier to review and merge. In this case, I would suggest having two PRs (one each for Lite and Micro).

I do apologize for asking you to start over but we truly are trying to reduce our response time with these new contribution guidelines.

I'm going to close this current PR and would encourage you to make a github issue to kickstart this conversation.

lamarrr commented:
> @lamarrr thanks for your patience.
> 
> Given the long history of this PR and the recent changes to the [Micro contribution guidelines](https://github.com/tensorflow/tensorflow/blob/master/tensorflow/lite/micro/CONTRIBUTING.md), I am going to request (with apologies for all the delays) that we take a step back and follow the new guidelines.
> 
> My specific recommendation is:
> 
> * Please make a github issue [motivating the new feature](https://github.com/tensorflow/tensorflow/blob/master/tensorflow/lite/micro/CONTRIBUTING.md) that you would like -- in an effort to give contributors faster feedback and PR reviews, we need to focus efforts on features that have broad impact. This will unfortunately mean saying no to some features.
> 
> If we agree that now is a good time for this feature, I would say the next steps would be:
> 
> * Since there was a lot of back and forth with @renjie-liu write up the details of the int8 math in the github issue so that is is documented for future reference and can also be referred to during the PR review.
> * Break the work into smaller PRs that will be easier to review and merge. In this case, I would suggest having two PRs (one each for Lite and Micro).
> 
> I do apologize for asking you to start over but we truly are trying to reduce our response time with these new contribution guidelines.
> 
> I'm going to close this current PR and would encourage you to make a github issue to kickstart this conversation.

Thanks! I understand.
I'll take the recommended steps.
